### PR TITLE
Refractoring kevin

### PR DESCRIPTION
**Pull request by TApplencourt**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- New feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Build related changes
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes
- Other (please describe):

### Does this introduce a breaking change?

- Yes
- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

**Review by ye-luo — CHANGES_REQUESTED** [review]
Must merge here after the PR got accepted upstream.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp [hunk 1]
```diff
@@ -97,27 +97,52 @@ void MultiDiracDeterminant::mw_BuildDotProductsAndCalculateRatios_impl(
...
+    for (size_t ext_level = 1; ext_level <= max_ext_level; ext_level++) {
+        count_0 += (*ndets_per_excitation_level_)[ext_level];
+        it_shift += (*ndets_per_excitation_level_)[ext_level] * (3 * ext_level + 1);
+        sum_ndets_per_excitation_level.push_back(count_0);
+        sum_with_shift_ndets_per_excitation_level.push_back(it_shift);
     }
-    count_0 += (*ndets_per_excitation_level_)[ext_level];
-    it_shift += (*ndets_per_excitation_level_)[ext_level] * (3 * ext_level + 1);
+  }
+  // Can replaced by variadic template at some point if needed
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp]: We probably need to choose different parallelization strategy for small and large max_ext_level.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp [hunk 2]
```diff
@@ -97,27 +97,52 @@ void MultiDiracDeterminant::mw_BuildDotProductsAndCalculateRatios_impl(
...
+  // Can put break to short-circuit
+  if (max_ext_level <= 0)
+    update_ratios_list<0>(nw, ratios_list,
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp]: We don't actually do case 0. max_ext_level starts from 1. See the old loop.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp [hunk 3]
```diff
@@ -55,7 +55,7 @@ void MultiDiracDeterminant::BuildDotProductsAndCalculateRatios_impl(int ref,
     const size_t n = *it2;
     //ratios[count]=(count!=ref)?sign[count]*det0*CalculateRatioFromMatrixElements(n,dotProducts,it2+1):det0;
     if (count != ref)
-      ratios[count] = sign[count] * det0 * CalculateRatioFromMatrixElements(n, dotProducts, it2 + 1);
+      ratios[count] = sign[count] * det0 * CalculateRatioFromMatrixElements(n, dotProducts, &(*(it2 + 1)) );
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp]: Please remove all `&(*`

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h [hunk 4]
```diff
@@ -243,20 +260,99 @@ class MultiDiracDeterminant : public WaveFunctionComponent
...
     default:
       return DetCalculator.evaluate(dotProducts, it, n);
     }
     return 0.0;
   }
 
+  // Anouar explain what does kevin did
+  //
+  // Do not offlaod this one
+  void update_ratios_list(int ext_level,
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h]: `mw_updateRatios`

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h [hunk 5]
```diff
@@ -243,20 +260,99 @@ class MultiDiracDeterminant : public WaveFunctionComponent
...
+            CalculateRatioFromMatrixElements(ext_level, dotProducts_list[iw].get(),
+                                             &(*(it2 + 1 + count_1 * (3 * ext_level + 1))));
+      }
+    }
+  }
+
+  // Anouar explain what does kevin did
+  // This one should be offloaded
+  template<int Ext_level>
+  void update_ratios_list(int nw,
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h]: `mw_updateRatios`

**Comment by ye-luo** [comment]
Also need to resolve the conflict.

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp (hunk 2 above)
Comment (TApplencourt) [reply src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp]: Will remove it!

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp (hunk 3 above)
Comment (TApplencourt) [reply src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.2.cpp]: @anbenali can you do it?  This one is to transform the iterator into a pointer. will need to change more thing for the offload :)

File: src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h (hunk 4 above)
Comment (TApplencourt) [reply src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.h]: Will change the name!

**Pull request closed without merging**